### ISSUE TITLE
[Feature Request] Sparse-Dense elementwise Multiplication 

### ISSUE THREAD

**Issue by chivee** [pytorch/pytorch#3158]
Hi, thanks again for the great work.

I would like to raise an issue to discussion how to implement the following features:
Dense[m,1] * Sparse[m,n] -> Sparse[m,n]
Sparse[m,n] * Dense[1,n] -> Sparse[m,n]
Sparse[m*n] * Dense[b,1,n] -> Sparse[b,m,n]

these features are fully supported by the [cuSPARSE Level 3](https://developer.nvidia.com/cusparse), as for the CPU side, I'm not quite sure weather MKL was satisfied. 

I would like to help on the implementation if needed, could anyone give me a brief guideline that what have be done and what still need to be do?

 

cc @ezyang @gchanan @zou3519 @bdhirsh @jbschlosser @aocsa @nikitaved @pearu @mruberry @vincentqb

**Comment by apaszke** [comment]
Just to clarify: are you looking for an elementwise product, or want to do matrix multiplication? From the shapes I infer it’s the first one, but I can’t find any related functions in cuSPARSE

**Comment by chivee** [comment]
I'm looking for element-wise product. I'm misunderstood that  [this fuction](http://docs.nvidia.com/cuda/cusparse/index.html#cusparse-lt-t-gt-csrmm) was describing an element-wise production. 

for `Dense[m,1] * Sparse[m,n] -> Sparse[m,n]` and `Sparse[m,n] * Dense[1,n] -> Sparse[m,n]` mainly have two issues:
         1. SparseTensor doesn't offer cmul(DenseTensor), which can be solved by convert Dense to Sparse first.
         2. Sparse * Sparse required the two tensor have [exactly same size ](https://github.com/pytorch/pytorch/blob/master/torch/lib/THCS/generic/THCSTensorMath.cu#L411). To my understanding if two tensor was broadcastable, that should also work?

and for `Sparse[m*n] * Dense[b,1,n] -> Sparse[b,m,n]`. If cuSPARSE didn't offer something like `bspaddcmul `I'm guessing I can only using an for loop on Dense[b,1,n] right?

**Comment by apaszke** [comment]
1. Converting dense tensors to sparse is a bad idea. It will take a lot more memory than the original dense tensor and will be extremely slow. We should write specialized kernels for this.
2. That's true, although I don't think our current broadcasting code supports sparse tensors. In general, broadcasting for sparse tensors is much harder and more expensive than for dense tensors, because you actually need to repeat all the values, making the tensor much larger

**Comment by chivee** [comment]
is there any plan that when should pytorch support CSR/CSC format? most of these operation will be cheap if using CSR/CSC.

for these special case, i guess we don't need really allocate more memories since every value should be repeated and only be used as a intermedia value? 

Besides, if the cuSPARSE and MKL didn't offer these implementation, does that mean that we need more effort to utilize the multiprocess?

**Comment by apaszke** [comment]
I don't think we'll be supporting other formats. They require new kernels for all math operations, and it's a lot of code to write and maintain.

That's true, but only if you special case the "broadcasted" operators in the backend. There's no way to easily "expand" the sparse tensor outside of a math method. It's because all non-zero values have to be listed explicitly, and expanding actually adds more non-zero values.

No, that's not true. We're using OpenMP in our code, which makes it very easy to paralellize math operations

**Comment by Mandeep-Rathee** [comment]
hey, 

I am also struggling with the same issue. I want to do element-wise multiplication of sparse and dense matrices with auto_grad = True. 
Did you find some solution for the same?

thanks in Advance.

**Comment by ladzin** [comment]
I ran into the same issue, even with the latest Pytorch build (1.8)

Simple example code:

A = torch.sparse_coo_tensor([[0,1],[0,1]], [1.0, 2.0])
B = torch.tensor([[2.0, 3.0]])
A*B # fails, but A.to_dense()*B works fine

**Comment by mruberry** [comment]
Bumping priority based on user activity.

**Comment by mruberry** [comment]
We would accept a PR implementing this behavior.

**Comment by aocsa** [comment]
I was looking this issue and seems feasible.  To focus on one goal, first I will implement the element-wise product between sparse tensor and dense tensor. 

cc @mruberry, @rgommers

**Comment by rgommers** [comment]
The issue description says that this should always result in sparse output. That sounds sensible, but there was a lot of discussion on the semantics of this elsewhere. @pearu is this actionable right now?

**Comment by pearu** [comment]
> The issue description says that this should always result in sparse output. That sounds sensible, but there was a lot of discussion on the semantics of this elsewhere. @pearu is this actionable right now?

It has been suggested [torch.sparse design principles] that "Binary functions operating on a COO tensor and a strided tensor generally produce strided tensors" that differs from the wished behavior indeed. So, the answer is no until the semantics of element-wise multiplication is confirmed.

A little background: While for element-wise multiplication, `COO * Strided -> COO` sounds sensible then for element-wise addition, `COO + Strided -> Strided` is inevitable. Hence the general suggestion for binary operations.

The  `COO * Strided -> COO` behavior could be achieved/implemented via `torch.mul(coo, strided, layout=torch.coo_layout)` or `torch.mul(coo, strided, out=out)` where `out` is a COO tensor with the same indices set as `coo` has.

**Comment by rgommers** [comment]
@mruberry this is blocked as well as lower-prio than other items on the sparse tracking issue (gh-44634), so I'll remove the high-prio label from this and add it to that roll-up.

**Comment by mruberry** [comment]
> @mruberry this is blocked as well as lower-prio than other items on the sparse tracking issue ([gh-44634](https://github.com/pytorch/pytorch/issues/44634)), so I'll remove the high-prio label from this and add it to that roll-up.

OK -- let's talk about this on Friday, too. I don't want us to feel blocked on a tracker. The tracker's supposed to help us!

**Comment by rgommers** [comment]
It's not blocked on the tracker, that's secondary here. It's blocked on "we don't yet all agree on the semantics", and resolving that is currently on the TODO list for after the CSR/GCS format work is landed.

**Comment by adityaradk** [comment]
Any information on when we can expect this functionality to be implemented?

**Comment by nikitaved** [comment]
Would be great to have, otherwise sparse feature selection is not possible.

**Comment by nikitaved** [comment]
This will be implemented once https://github.com/pytorch/pytorch/pull/81556 is merged.

**Comment by plutonium-239** [comment]
The PR mentioned was reverted in August, and there have been no updates since... Is it being worked on?

**Comment by nikitaved** [comment]
@plutonium-239 , it is in PyTorch since 1.13.

**Comment by plutonium-239** [comment]
Ah! Just need to update from 1.8 then 😅
thanks!
